contracts/SimpleSaleContract.sol

```diff
@@ -16,47 +23,140 @@ contract SimpleSaleContract is Ownable {
      * @param _paymentToken address of token to be set
-     * @param _exchangeRate exchange rate of the token in relation to mintable token
+     * @param _partsSellToken parts of the token that is sold to be converted into _partsMintToken
+     * @param _partsMintToken parts of the token that is minted in exchage of _partsSellToken
      */
-    function setPaymentTokenExchangeRate(address _paymentToken, uint256 _exchangeRate) external onlyOwner {
-        require(_paymentToken != address(0), "Payment token not set");
+    function setPaymentTokenExchangeRate(IERC20 _paymentToken, uint256 _partsSellToken, uint256 _partsMintToken) external onlyOwner {
+        require(address(_paymentToken) != address(0), "Payment token is invalid");
+        require(_partsSellToken != 0, "Invalid exchange rate");
+        require(_partsMintToken != 0, "Invalid exchange rate");
 
-        paymentTokensExchangeRate[IERC20(_paymentToken)] = _exchangeRate;
+        paymentTokensExchangeRate[_paymentToken].partsSell = _partsSellToken;
+        paymentTokensExchangeRate[_paymentToken].partsMint = _partsMintToken;
+
+        emit TokenExchangeRateSet(_paymentToken, _partsSellToken, _partsMintToken);
+    }
+
+    /**
+     * @notice admin function for unsetting payment token exchange rate
+     * @param _paymentToken address of token to be unset
+     */
+    function unsetPaymentTokenExchangeRate(IERC20 _paymentToken) external onlyOwner {
+        // Is it cheaper to call delete than x.a=0, x.b=0?
+        delete paymentTokensExchangeRate[_paymentToken];
+    }
+
+    /**
+     * @notice admin function to withdraw contract balance of a certain token
+     * @param _token address of a token
+     */
+    function withdraw(IERC20 _token) external onlyOwner {
+        require(block.timestamp >= cutoffTimestamp, "Sale is still ongoing");
+
+        uint256 balance = _token.balanceOf(address(this));
+        _token.safeTransfer(msg.sender, balance);
+
+        emit TokensWithdrawn(_token, balance);
+    }
+
+    /**
+     * @notice Transfers mintableTokens to msg.sender for a specified exchange rate of ether
+     * @param _tokensToBuy number of mintable tokens to be bought
+     */
+    function buyTokensForEther(uint256 _tokensToBuy) external payable forSale positiveTokens(_tokensToBuy) rateSet(WETH) {
+        uint256 etherToPay = _tokensToBuy 
+            * paymentTokensExchangeRate[WETH].partsSell
+            / paymentTokensExchangeRate[WETH].partsMint;
+        
+        if (msg.value != etherToPay) {
+            revert BadEthValue();
+        }
 
-        emit TokenExchangeRateSet(_paymentToken, _exchangeRate);
+        // This doesn't make sense because we are approving a transfer to ourselves o_O
+        // Should we remove the _transferFrom() part from the _buyTokens 
+        // And only use the _mint() part?
+        Weth(address(WETH)).deposit{value: msg.value}();
+        WETH.approve(address(this), msg.value);
```

Comment:
Are you sure that this is working correctly? You do an approve but the msg.sender is the sale contract. So it's quite unnecessary to do self approve

Reply:
not using _buyTokens() anymore